beacon_chain/validator_induction/src/inductor.rs

```diff
@@ -1,278 +1,187 @@
+use spec::ChainSpec;
 
 #[derive(Debug, PartialEq, Clone)]
 pub enum ValidatorInductionError {
     InvalidShard,
     InvaidProofOfPossession,
+    InvalidWithdrawalCredentials
 }
 
-impl ValidatorInductor {
-    pub fn new(current_slot: u64, shard_count: u64, validators: Vec<ValidatorRecord>) -> Self {
-        Self {
-            current_slot,
-            shard_count,
-            validators,
-            empty_validator_start: 0,
-        }
-    }
-
-    /// Attempt to induct a validator into the CrystallizedState.
-    ///
-    /// Returns an error if the registration is invalid, otherwise returns the index of the
-    /// validator in `CrystallizedState.validators`.
-    pub fn induct(
-        &mut self,
-        rego: &ValidatorRegistration,
-        status: ValidatorStatus,
-    ) -> Result<usize, ValidatorInductionError> {
-        let v = self.process_registration(rego, status)?;
-        Ok(self.add_validator(v))
-    }
-
-    /// Verify a `ValidatorRegistration` and return a `ValidatorRecord` if valid.
-    fn process_registration(
-        &self,
-        r: &ValidatorRegistration,
-        status: ValidatorStatus,
-    ) -> Result<ValidatorRecord, ValidatorInductionError> {
-        /*
-         * Ensure withdrawal shard is not too high.
-         */
-        if r.withdrawal_shard > self.shard_count {
-            return Err(ValidatorInductionError::InvalidShard);
-        }
-
-        /*
-         * Prove validator has knowledge of their secret key.
-         */
-        if !verify_proof_of_possession(&r.proof_of_possession, &r.pubkey) {
-            return Err(ValidatorInductionError::InvaidProofOfPossession);
-        }
-
-        Ok(ValidatorRecord {
-            pubkey: r.pubkey.clone(),
-            withdrawal_shard: r.withdrawal_shard,
-            withdrawal_address: r.withdrawal_address,
-            randao_commitment: r.randao_commitment,
-            randao_last_change: self.current_slot,
-            balance: DEPOSIT_GWEI,
-            status: status,
-            exit_slot: 0,
-        })
-    }
-
-    /// Returns the index of the first `ValidatorRecord` in the `CrystallizedState` where
-    /// `validator.status == Withdrawn`. If no such record exists, `None` is returned.
-    fn first_withdrawn_validator(&mut self) -> Option<usize> {
-        for i in self.empty_validator_start..self.validators.len() {
-            if self.validators[i].status == ValidatorStatus::Withdrawn {
-                self.empty_validator_start = i + 1;
-                return Some(i);
-            }
-        }
-        None
+pub fn process_deposit(
+    state: &mut BeaconState,
+    deposit: &Deposit,
+    spec: &ChainSpec) 
+-> Result<usize, ValidatorInductionError> {
+    let deposit_input = &deposit.deposit_data.deposit_input;
+    let deposit_data = &deposit.deposit_data;
+        
+    if !verify_proof_of_possession(&deposit_input.proof_of_possession, &deposit_input.pubkey) {
```

Comment:
I wasn't able to figure out how this should look based on the spec. If it's not a big deal, it might be best to update this in another PR.

Reply:
Sorry, which part are you unsure about?

Reply:
I wasn't sure how implement the `validate_proof_of_possession` helper method as defined in the [spec](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#routine-for-processing-deposits), partly because there is no `domain` parameter in our BLS signature, and also because there doesn't seem to be an implementation of `hash_tree_root`. Apologies for the lack of clarity. It looks to me like this part of the deposit function should get refactored at a later point.

Reply:
Oh yep, I'm with you now.

I agree it'll need a refactor. I think it's fine to leave it like this as long as we drop a `TODO` in there referencing #91 and #70 :)